File: Server/Serveur/Manager/UserManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Serveur
{
    class UserManager
    {
        Dictionary<int, User> existingUsers; // int = userId

        public UserManager()
        {
            existingUsers = new Dictionary<int, User>();
        }

        public void Initialize()
        {
            // Build existingUsers from file parser
        }

        public bool AddNewUser(String userName)
        {
            foreach(User u in GetListExistingUsers())
            {
                if (u.GetName() == userName)
                    return false;
            }

            existingUsers.Add(existingUsers.Count, new User(existingUsers.Count, userName));
            return true;
        }

        public bool SendInvitation(int userId, int groupId)
        {
            if (!existingUsers.ContainsKey(userId))
                return false;

            existingUsers[userId].AddPendingInvitation(groupId);
            return true;
        }

        public List<int> GetAllGroupIdFor(int id)
        {
            if (existingUsers.ContainsKey(id))
                return existingUsers[id].getGroups();

            return null;
        }

        public void Synchronize()
        {
            foreach (User u in existingUsers.Values.ToList())
            {
                u.Synchronize();
            }
        }

        public List<User> GetListExistingUsers()
        {
            return existingUsers.Values.ToList();
        }

        public Dictionary<int, User> GetExistingUsers()
        {
            return existingUsers;
        }

        public User GetUserByName(string userName)
        {
            User user = null;

            foreach(User u in GetListExistingUsers())
            {
                if(u.GetName() == userName)
                {
                    user = u;
                    break;
                }
            }

            return user;
        }

        public User GetUserById(int userId)
        {
            if (!DoesUserExists(userId))
                return null;

            return existingUsers[userId];
        }

        private bool DoesUserExists(int userId)
        {
            return existingUsers.ContainsKey(userId);
        }
    }
}

File: Server/Serveur/UserManager.cpp
#include "UserManager.h"

UserManager* UserManager::um = nullptr;

UserManager::UserManager()
{
    existingUsers = map<unsigned int, User>();
}

UserManager* UserManager::getInstance() {
	if (UserManager::um == nullptr)
		UserManager::um = new UserManager();

	return UserManager::um;
}

bool UserManager::addNewUser(string userName) {
    for(auto u : getListExistingUsers())
    {
        if (u.getName() == userName)
            return false;
    }

    existingUsers.insert(make_pair(existingUsers.size(), User{ existingUsers.size(), userName }));
    return true;
}

vector<User> UserManager::getListExistingUsers() {
    vector<User> users;

    for (auto u : existingUsers) {
        users.push_back(u.second);
    }

    return users;
}

User UserManager::getUserByName(string userName) {
    User user;

    for(auto u : getListExistingUsers())
    {
        if (u.getName() == userName)
        {
            user = u;
            break;
        }
    }

    return user;
}

User UserManager::getUserById(unsigned int userId) {
    if (!doesUserExists(userId))
        throw exception{"No user exist for this id"};

    return existingUsers[userId];
}

bool UserManager::doesUserExists(unsigned int userId) {
    return existingUsers.count(userId);
}